### PR TITLE
Reorganize web console create flow

### LINKED ISSUE #4331: No way to set --context-dir when creating new app via web console
Env:
----
~~~
$ openshift version
openshift v1.0.4-491-g5193c6b
kubernetes v1.1.0-alpha.0-1605-g44c91b1
~~~

Issue:
---
I can create new application with `--context-dir` from CLI like:
~~~
$ oc new-app https://github.com/openshift/sti-php.git --context-dir=5.5/test/test-app -l app=php --strategy=source
~~~

But it looks like there are no way to set `--context-dir` via web console.

context-dir-report

### PR DESCRIPTION
**Pull request by spadgett**
Present builder images and templates in one catalog grouped by tag.

- [x] Update catalog layout
- [x] Show helpful message on empty catalog
- [x] Add show more link when category has many entries
- [x] Make tag check case insensitive
- [x] Filter by keyword and tag
- [x] Make tag text in tiles clickable filters
- [x] Use sample repo for selected builder
- [x] Autofocus name field on create from image page
- [x] Change `imageRepo` var names to `imageStream`
- [x] Update empty project message for new create options
- [x] Go directly to create page after adding a new project
- [x] Move most items into advanced options section to create from image page
- [x] Search for keywords in description
- [x] Add back git ref URL fragment
- [x] Look for sample git ref and context dir annotations
- [x] Fix tag styles for image stream page
- [x] Fix problem with mixed case tags
- [x] Update placeholder text in create form
- [x] https://bugzilla.redhat.com/show_bug.cgi?id=1266390
- [x] https://bugzilla.redhat.com/show_bug.cgi?id=1266377

Fixes #4331

### CODE REVIEW

**Comment by spadgett** [comment]
Create from image page updates, using the sample repos from #4619.

openshift_web_console

@jwforres

**Comment by spadgett** [comment]
Empty state message. @jwforres @ajacobs21e opinions?

openshift_web_console

**Comment by spadgett** [comment]
@ajacobs21e is there a standard style or icon for back links?

**Comment by ajacobs21e** [comment]
@spadgett where does the back link go? To overview? Would it make more sense to say "Back to project overview" instead? The link style is right. 

The messaging looks good. I would remove the comma after "run the command" or use a colon instead.

**Comment by spadgett** [comment]
Thanks, @ajacobs21e. I made your suggested changes.

**Comment by jwforres** [comment]
@spadgett also keep in mind really soon that we are going to let you upload your own template to use from this page. so this message will probably change a bit

cc @gruiz17

**Comment by spadgett** [comment]
Added filtering by keyword.

openshift_web_console

**Comment by ajacobs21e** [comment]
+1

On Wed, Sep 16, 2015 at 2:12 PM, Sam Padgett <notifications@github.com>
wrote:

> Added filtering by keyword.
>
> [image: openshift_web_console]
> <https://cloud.githubusercontent.com/assets/1167259/9913865/f45a8d90-5c7c-11e5-8c9a-d7ab34569844.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4684#issuecomment-140826294>.
>



-- 

Allie Jacobs
UXD

calendar
<https://www.google.com/calendar/b/1/embed?src=ajacobs@redhat.com&ctz=America/New_York>

**Comment by spadgett** [comment]
@jwforres Change the description to a tooltip like v2? So little fits on the page. Even switching to 2 columns, you'd only see 6 in this screenshot.

openshift_web_console_and_rails-postgresql_json____go_src_github_com_openshift_origin_tmp_files_examples_quickstart-templates__-_vim_and_more_template_changes_-_spadgett_redhat_com_-_red_hat_mail

**Comment by spadgett** [comment]
Without descriptions... @sg00dwin is looking at keeping a group in one column like v2.

screen shot 2015-09-16 at 4 14 54 pm

**Comment by ajacobs21e** [comment]
When you select one does it still open a modal or go directly to the next
step?

On Wed, Sep 16, 2015 at 4:16 PM, Sam Padgett <notifications@github.com>
wrote:

> Without descriptions... @sg00dwin <https://github.com/sg00dwin> is
> looking at keeping a group in one column like v2.
>
> [image: screen shot 2015-09-16 at 4 14 54 pm]
> <https://cloud.githubusercontent.com/assets/1167259/9917175/2b4663a4-5c8e-11e5-88fe-83dc720607f9.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4684#issuecomment-140873583>.
>



-- 

Allie Jacobs
UXD

calendar
<https://www.google.com/calendar/b/1/embed?src=ajacobs@redhat.com&ctz=America/New_York>

**Comment by spadgett** [comment]
It goes to the next step, but the description is in the next step and it's easy to go back. It's also in a tooltip.

**Comment by jwforres** [comment]
Maybe we should use a bootstrap popover instead of browser title text? That
way the tooltip is immediate on hover?  Browser title hovers aren't the
most discoverable things ever :)

**Comment by spadgett** [comment]
I think that would be better. I'm not sure if the bootstrap popovers work with angular expressions, though.

**Comment by spadgett** [comment]
Would it be distracting when you move your cursor around the page if the popovers appeared immediately?

**Comment by spadgett** [comment]
I assume the hover target would be an icon and not the whole tile?

**Comment by jwforres** [comment]
See the pod-warnings directive for a dynamic popover example.  Also, we may
want to have the popover be from a ? icon if we find the popovers are
getting in the way of clicking on other tiles.

**Comment by spadgett** [comment]
Basic support for browse by tag.

openshift_web_console

**Comment by ajacobs21e** [comment]
Is the reason we are considering a column layout like v2 because categories may have less than 3 items? The layout with section headers on the left is easier for scanning but could lead to wasted space.

**Comment by spadgett** [comment]
> Is the reason we are considering a column layout like v2 because categories may have less than 3 items?

It looked really bad to me when several groups had only one item, almost all whitespace.

**Comment by ajacobs21e** [comment]
Ah ok that makes sense, thanks

On Thu, Sep 17, 2015 at 2:05 PM, Sam Padgett <notifications@github.com>
wrote:

> Is the reason we are considering a column layout like v2 because
> categories may have less than 3 items?
>
> It looked really bad to me when several groups had only one item, almost
> all whitespace.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4684#issuecomment-141168827>.
>



-- 

Allie Jacobs
UXD

calendar
<https://www.google.com/calendar/b/1/embed?src=ajacobs@redhat.com&ctz=America/New_York>

**Comment by spadgett** [comment]
Here is how things look with columns. Some cleanup left to do. Thanks @sg00dwin.

screen shot 2015-09-17 at 3 32 27 pm

**Comment by liggitt** [comment]
Definitely like one column. No descriptions seems terse

**Comment by spadgett** [comment]
Thoughts on adding the cube icon for builder images?

openshift_web_console

**Comment by ajacobs21e** [comment]
https://docs.google.com/document/d/1p_3Yh4Wxu2y4V7oKavxLf5pc7Evu0W61qBmv169QtGg/edit
That icon is going to be used for "container" so we'd have to find another icon. I like the concept though.

**Comment by spadgett** [comment]
@ajacobs21e Good to know. We're using `fa-cube` for image in other places in the UI currently, so I think we should fix them all when we have a new image icon.

**Comment by ajacobs21e** [comment]
@spadgett Yeah that's mentioned in the doc as a conflict. There's a proposed new image icon and I'll have to track down the status of that. The containers UI sharing meeting has been canceled the past few weeks. I'm adding these questions to the agenda now so hopefully I can get answers Tuesday.

**Comment by spadgett** [comment]
From web console design review feedback -- thoughts on hiding most options under an advanced options link?

@sspeiche fyi

screen shot 2015-09-18 at 4 38 28 pm

**Comment by liggitt** [comment]
I like hiding under a reveal link... I would be more descriptive in the link, e.g. "Build from a different branch or subdirectory of your source repository"

**Comment by spadgett** [comment]
It's everything else under there: routing, deployment config, build config, replicas, labels...

**Comment by sspeiche** [comment]
I think it is a nice improvement.  It would be good to look at a holistic view of it.  What I mean is, can we make it so the user doesn't have to enter any data? and evaluable how much visual space just the 2 fields for name and repo URL take.

It feels like the whole name entry is way too much screen real estate for something that is obvious and no one will probably change. Perhaps it could be put in the title/header and made editable. I know it is important for how other resources are named, though I wonder how many users (especially new users) need to be educated about this. Perhaps a hover or a ? to get more info about the name.

Like the Git repo URL seems like a lot of space too. It would be nice if we had a greyed text of the sample URL itself in the box and we just used it if they didn't change it.

So in other words, if I went to this page I could just hit Submit without filling in anything.

Agree with Mr @liggitt that being more task focused than just "advanced" would be helpful.

There is my 2 ¢

**Comment by liggitt** [comment]
I don't want to default the source URL… they need to choose to use a sample or set their own. Anything that important shouldn't be set so you can accidentally miss it and complete the wizard

**Comment by liggitt** [comment]
@spadgett if you get autofocus working, apply the same technique to the new project page

**Comment by sspeiche** [comment]
Defaulting would probably not be a good idea, unless there was an easy way to change after creation (that would be handy)

**Comment by spadgett** [comment]
I feel like name should be very prominent. It's the name of your app, and "php" or "wildfly" isn't a great choice. It's not easy to change later (is it even possible?). I wonder if we should remove the prefill if we're not using the Git URL to suggest a name like before.

**Comment by spadgett** [comment]
[test]

**Comment by jwforres** [comment]
@spadgett the reason we pre-filled in v2 was so you could always go to the create page and just hit submit.  With this you already have to tell us what source URL you want so its not that much worse to have to pick a useful name too.

honestly i think we pre-filled it in v2 because we wanted to be lazy when we were testing, not because it was all that beneficial to a new user... but maybe that's just me :)

**Comment by liggitt** [comment]
We only pre-filled in v2 when creating from quickstarts which defined a cartridge and a source repo, right?

**Comment by spadgett** [comment]
It's pre-filling for me in v2 no matter what I pick.

**Comment by liggitt** [comment]
yeah, I thought you were talking about the git repo, not the name

**Comment by spadgett** [comment]
@liggitt 

> if you get autofocus working, apply the same technique to the new project page

Fixed in both places.

File: assets/app/scripts/controllers/create.js [hunk 1]
```diff
@@ -47,33 +119,198 @@ angular.module('openshiftConsole')
...
-      $scope.templatesLoaded = openshiftTemplates && projectTemplates;
-      $scope.instantApps = {};
-      $scope.otherTemplates = {};
-
-      // Categorize templates as instant apps or "other."
-      var categorizeTemplates = function(template) {
-        var uid = uidFilter(template);
-        if (isInstantApp(template)) {
-          $scope.instantApps[uid] = template;
+    function matchesKeyword(name, tags, keyword) {
```
Comment (jwforres) [inline assets/app/scripts/controllers/create.js]: so we are only matching keyword against the name or the tags?  did you consider also matching on the description

Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: Yes, and I can go either way on this. It resulted in some unexpected matches. For instance, searching for "database" matched templates with descriptions like "An example application with no database." Maybe that's OK.

It's a little odd as well if there's only a match in the description because we don't show the description. It's not clear why it's a match.

File: assets/app/scripts/controllers/create.js [hunk 2]
```diff
@@ -47,33 +119,198 @@ angular.module('openshiftConsole')
...
+    function updateColumns() {
+      var i,
+          numLeft = 0,
+          numRight = 0,
+          categories = $scope.filteredCategoryTags;
+
+      $scope.leftCategories  = [];
+      $scope.rightCategories = [];
+
+      for (i = 0; i < categories.length; i++) {
```
Comment (jwforres) [inline assets/app/scripts/controllers/create.js]: either angular.forEach or var category = categories[i] so you dont have to repeat categories[i]

File: assets/app/scripts/controllers/create.js (hunk 1 above)
Comment (sspeiche) [inline assets/app/scripts/controllers/create.js]: I like the idea of searching description too (if even an option, which sounds complicated).

Perhaps we(I) could help create that "guidance" document around writing templates so they look good in Web UI. Could add this guidance when writing description to remove this false positive hits.

Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: I'll add description search.

File: assets/app/scripts/controllers/create.js [hunk 3]
```diff
@@ -47,33 +119,198 @@ angular.module('openshiftConsole')
...
+            imageStream: imageStream,
+            imageStreamTag: imageStreamTag,
+            name: imageStream.metadata.name + ":" + imageStreamTag,
+            version: imageStreamTagAnnotationFilter(imageStream, 'version', imageStreamTag),
+            categoryTags: categoryTags
+          };
+          if (categoryTags.indexOf("builder") >= 0) {
+            // Add the builder image to its category.
+            category = getCategory(categoryTags);
+            if (!buildersByCategory[category]) {
```
Comment (jwforres) [inline assets/app/scripts/controllers/create.js]: we usually make this a one-liner `buildersByCategory[category] = buildersByCategory[category] || [];`

File: assets/app/scripts/controllers/create.js [hunk 4]
```diff
@@ -47,33 +119,198 @@ angular.module('openshiftConsole')
...
 
-    $scope.createFromSource = function() {
-      if($scope.from_source_form.$valid) {
-        var createFromSourceURL = createFromSourceURLFilter($scope.projectName, $scope.from_source_url);
-        $location.url(createFromSourceURL);
+    function categorizeTemplates(templates) {
+      angular.forEach(templates, function(template) {
+        var tags = tagsFilter(template);
+        var category = getCategory(tags);
+        if (!templatesByCategory[category]) {
```
Comment (jwforres) [inline assets/app/scripts/controllers/create.js]: `templatesByCategory[category] = templatesByCategory[category] || [];`

File: assets/app/scripts/controllers/create.js [hunk 5]
```diff
@@ -47,33 +119,198 @@ angular.module('openshiftConsole')
...
+      for (i = 0; i < $scope.categoryTags.length; i++) {
+        for (j = 0; j < tags.length; j++) {
+          if (tags[j].toLowerCase() === $scope.categoryTags[i]) {
+            return tags[j];
+          }
+        }
       }
-    };
+
+      return "";
```
Comment (jwforres) [inline assets/app/scripts/controllers/create.js]: add a comment here that says this means it goes into the "other" category so someone doesnt try to change it to null or something later

File: assets/app/scripts/controllers/create.js [hunk 6]
```diff
@@ -47,33 +119,198 @@ angular.module('openshiftConsole')
...
+    }
+
+    function updateState() {
+      // Have we finished loading all of the templates and image streams in
+      // both the project and openshift namespaces? If undefined, they're not
+      // loaded.
+      $scope.loaded =
+        projectTemplates &&
```
Comment (jwforres) [inline assets/app/scripts/controllers/create.js]: what if some of the requests succeed and some fail, on failures do you set these to empty maps?

**Comment by spadgett** [comment]
Added description to keyword search.

File: assets/app/scripts/services/applicationGenerator.js [hunk 7]
```diff
@@ -231,8 +222,8 @@ angular.module("openshiftConsole")
           },
           source: {
             git: {
-              ref: sourceRef,
-              uri: sourceUrl
+              ref: input.buildConfig.gitRef || 'master',
```
Comment (jwforres) [inline assets/app/scripts/services/applicationGenerator.js]: v2 actually allowed specifying the ref in either the URL with a # or by putting it in the separate ref input field, so that you could use the same "URL" in both the CLI and the web.  Its a bit weird I know.

File: assets/app/scripts/controllers/create.js (hunk 6 above)
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: I'm not sure we want to. We'll show any data that's there. If we set things to empty maps, it might result in an incorrect message saying you have no images or templates. You do get an error alert from DataService.

File: assets/app/scripts/services/applicationGenerator.js (hunk 7 above)
Comment (spadgett) [inline assets/app/scripts/services/applicationGenerator.js]: I'll add it back. What if git ref is both in the URL and a different value is specified in the input?

Comment (liggitt) [inline assets/app/scripts/services/applicationGenerator.js]: if they paste in something with a fragment, remove it and populate the ref field with it if empty?

File: assets/app/views/catalog/_catalog-category.html [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  <div class="catalog">
+    <!-- Show builder images first. -->
+    <catalog-image
+      image-stream="builder.imageStream"
+      image-tag="builder.imageStreamTag"
+      project="{{project}}"
+      version="builder.version"
+      filter-tag="filterTag"
+      is-builder="true"
+      ng-repeat="builder in builders | orderBy : ['name', 'imageStream.metadata.namespace'] | limitToOrAll: itemLimit track by builderID(builder)">
```
Comment (jwforres) [inline assets/app/views/catalog/_catalog-category.html]: does imageStream.metadata.namespace ir orderBy work? were you able to try it?

Comment (spadgett) [inline assets/app/views/catalog/_catalog-category.html]: Yes, it works.

File: assets/app/views/create.html [hunk 9]
```diff
@@ -7,67 +7,117 @@
...
+          </div>
+
+          <!-- Show an abbreviated message if only non-builders exist in the project.
+               We show a link to view the non-builder images below. -->
+          <p ng-if="emptyCatalog && loaded && nonBuilderImages.length">No builder images or templates.</p>
+
+          <div ng-show="!emptyCatalog">
+            <h1>Select Image or Template</h1>
+
+            <p>Choose from web frameworks, databases, and other components.
```
Comment (jwforres) [inline assets/app/views/create.html]: Maybe "To get started, choose from these web frameworks, databases, and other components."

Just not sure about the second sentence "You'll be prompted... etc"

File: assets/app/views/create.html [hunk 10]
```diff
@@ -7,67 +7,117 @@
...
+              <div ng-if="filter.tag" class="filter-msg">
+                Tagged with {{filter.tag}}.
+                <a href="" ng-click="filter.tag = ''">See all tags</a>
               </div>
-              <div ng-show="!showAllTemplates && (otherTemplates | hashSize) !== 0" class="show-all-button-container gutter-bottom">
-                <a ng-click="showAllTemplates = true" class="btn btn-lg btn-primary">Show All Templates</a>
+            </div>
+
+            <div ng-if="filteredCategoryTags.length === 0" style="margin-top: 5px;">
+              All builder images and templates hidden by filter.
```
Comment (jwforres) [inline assets/app/views/create.html]: hidden by filter sounds weird, maybe hidden by the current filters.

File: assets/app/views/create.html (hunk 9 above)
Comment (spadgett) [inline assets/app/views/create.html]: Yeah, the "You'll be prompted..." was there before. I can take it out.

We use "Get started..." on the empty project page. I don't think we should repeat it again.

File: assets/app/views/create/fromimage.html [hunk 11]
```diff
@@ -14,179 +14,194 @@
...
-                  <div ng-show="$parent.expand">
-                    <div class="form-group">
-                      <label class="checkbox-inline">
-                        <input type="checkbox" ng-model="routing.include">
-                          Create a route to the application
-                          <span class="help action-inline">
-                            <i class="pficon pficon-help" data-toggle="tooltip" data-placement="right" data-original-title="Additional routes with alternative configuration options (security, TLS termination, etc) can be added using the oc command."></i>
-                          </span>
-                      </label>
+                      <input id="gitref" ng-model="buildConfig.gitRef" type="text"
```
Comment (jwforres) [inline assets/app/views/create/fromimage.html]: there can (and will for some jboss images) be a sample git ref annotation

File: assets/app/views/create/fromimage.html [hunk 12]
```diff
@@ -14,179 +14,194 @@
...
-                  expand="true"
-                  can-toggle="false"
-                  >
-                  <div ng-hide="$parent.expand">
-                    <h3>Autodeploy when</h3>
-                    <div>
-                      <label>New image is available: </label>
-                      <span>{{deploymentConfig.deployOnNewImage | yesNo}}</span>
-                    </div>
+                  <div class="form-group">
```
Comment (jwforres) [inline assets/app/views/create/fromimage.html]: there can (and will for some jboss images) be a sample git context dir annotation

File: assets/app/views/create/fromimage.html (hunk 11 above)
Comment (spadgett) [inline assets/app/views/create/fromimage.html]: I'll add that... What are the annotations for git ref and context dir?

**Comment by spadgett** [comment]
Updated some of the wording.

openshift_web_console

**Comment by jwforres** [comment]
Works for me.

Also, that's all my review comments.  You can check with @bparees for what the sample git ref and context dir annotations are.

File: assets/app/scripts/services/applicationGenerator.js (hunk 7 above)
Comment (spadgett) [inline assets/app/scripts/services/applicationGenerator.js]: Might be weird to have part of the URL disappear after the field blurs, particularly if the ref field is hidden under the advanced options link.

**Comment by spadgett** [comment]
@jwforres we should add git ref to the build config page. It's not there.

**Comment by jwforres** [comment]
Or context dir I bet, will add it to my list

File: assets/app/scripts/services/applicationGenerator.js [hunk 13]
```diff
@@ -205,16 +205,7 @@ angular.module("openshiftConsole")
         });
       }
 
-      // User can input a URL that contains a ref
-      var uri = new URI(input.buildConfig.sourceUrl);
-      var sourceRef = uri.fragment();
-      if (!sourceRef || sourceRef.length === 0) {
```
Comment (spadgett) [inline assets/app/scripts/services/applicationGenerator.js]: @jwforres Any reason the `sourceRef.length` check was needed here? Does `fragment()` not return a string?

Comment (jwforres) [inline assets/app/scripts/services/applicationGenerator.js]: Not needed, I just forgot "" is falsey, I can't ever keep that one straight
in my head.

**Comment by jwforres** [comment]
@spadgett the annotation names are here https://github.com/openshift/origin/pull/4619#issuecomment-139624158

**Comment by spadgett** [comment]
I kept one try it link that fills in everything.

openshift_web_console

**Comment by spadgett** [comment]
@jwforres I think that's everything so far.

**Comment by liggitt** [comment]
placeholder "master" for the git reference, and "/" for the context dir?

**Comment by spadgett** [comment]
Added. Is it strange to use default values for some placeholders and help text for others? I'm not sure "Repository URL" is helpful.

openshift_web_console

**Comment by spadgett** [comment]
Without repository URL placeholder:

openshift_web_console

**Comment by jwforres** [comment]
@spadgett yeah agree the repository URL placeholder was a bit redundant

**Comment by spadgett** [comment]
> yeah agree the repository URL placeholder was a bit redundant

Removed.

**Comment by spadgett** [comment]
https://ci.openshift.redhat.com/jenkins/job/fork_ami/89/

**Comment by nak3** [comment]
Is it impossible to create new-app without imagestream/template import?
For example, in case that I would like to create new-app by this style via webconsole:
`oc new-app someimage~http://somephprepo --strategy=source`

**Comment by jwforres** [comment]
Currently, here is the trello card tracking that
https://trello.com/c/fonu3qc5

On Thu, Sep 24, 2015 at 10:22 PM, Kenjiro Nakayama <notifications@github.com

**Comment by nak3** [comment]
Thank you! It will not contain this PR, but it seems that it will implement soon (via another PR(?)). That's great.

**Comment by jwforres** [comment]
Right, it won't be in this PR.  I'm not sure how soon we'll get to it,
there are a lot of things we want to add to the console :)

I suspect we will end up splitting that card into several smaller stories
and do a few of those use cases before we do others.

On Fri, Sep 25, 2015 at 11:01 AM, Kenjiro Nakayama <notifications@github.com

**Comment by spadgett** [comment]
Rebased.

**Comment by spadgett** [comment]
Rebased again.

**Comment by jwforres** [comment]
[merge]
